### DISCUSSION TITLE
Missing (editable) Imports: But python REPL has no problems to import

### DISCUSSION THREAD

**Discussion by buhtz** [microsoft/pylance-release#6384] [Enhancement]
## Environment data

-   Pylance version: v2024.9.1
-   OS and version: Windows 10 (I am forced to)
-   Python version (& distribution if applicable, e.g. Anaconda): Regular python.org Python "'3.11.5 (tags/v3.11.5:cce6ba9, Aug 24 2023, 14:38:34) [MSC v.1936 64 bit (AMD64)]"
- VSCode 1.93.0

## Problem

PyLance reports `Import "buhtzology" could not be resolved...` but regular python has no problem with it. This works fine and without errors:

    python -c "import buhtzology"

## Expected behavior

PyLance should use the same "environment" as the regular python interpreter does.

## Idea

It might be the case that PyLance is not able to handle packages installed as "--editable" (via python -m pip)
The package install data can be found in the path `C:\python\Lib\site-packages\buhtzology-0.2.0.dev0.dist-info`. Inside that folder the content of fiule `direct_url.json` points to the real python file:

    {"dir_info": {"editable": true}, "url": "file:///C:/Users/buhtzch/goodCloud/my.work/buhtzology"}

The Python interpreter can handle that without problems. So PyLance should also.

**Comment by buhtz** [comment]
Might be related to:
 - #78
 - #3473

**Comment by debonte** [comment]
Please see https://microsoft.github.io/pyright/#/import-resolution?id=editable-installs. If the `.pth` file for `buhtzology` contains only a path (no Python code), it will work with Pylance.

**Comment by buhtz** [comment]
That link does not explain anything and also does not offer a solution.
PyLance is the problem not the settings. The python interpreter (and a lot of other related tools like ruff, pylint, pycodestyle, pytest, codespell, ....) itself does not have a problem using that editable-installed packages. So why does PyLance have a problem with it?

**Comment by rchiodo** [comment]
The key part of that explanation is the 'static' part. Pylance is a static analyzer. It doesn't run python code. Dynamic code in .pth files cannot be evaluated in order to find imports. That's why it doesn't work. 

If you use compat mode, where it writes a file path out to the .pth file, that should work.

**Comment by buhtz** [comment]
So pylance parse python code as text file but never runs it?
PyLint also is a static linter but is able to handle editable installed packages.

**Comment by rchiodo** [comment]
I can't say how Pylint handles editable installs, but Pylance does not run any python code. It's not even written in Python.

**Comment by buhtz** [comment]
> I can't say how Pylint handles editable installs

Maybe then Pylint would be a good start to investigate a solution for PyLance.

I twas not clear from the previous Issues and not from docs. Do the maintainers of PyLance treat this as a real bug or is this intended behavior? When the "bug" is fixed doesn't matter at this question. Is it a bug: Yes or No?

**Comment by rchiodo** [comment]
As of right now this issue is by design. 

[Pylint](https://github.com/pylint-dev/pylint/tree/main/pylint) is actually written in Python, so likely can use the import mechanism in the runtime to attempt to import packages. Pyright/Pylance are written in typescript and have to model anything the python runtime does in typescript. 

In the past editable installs have always been answered with use compat mode, we don't support dynamic imports, this is by design. 

The problem is a lot of people do editable installs so maybe it's time we support some subset of dynamic imports. The problem is that ends up exploding into essentially emulating everything the python runtime does, so we end up back with just running python to find imports. 

I'm going to convert this item into a discussion for up votes.

**Reply by erictraut** [reply]
I think the right answer here is to work with the packaging community to come up with a declarative mechanism for editable installs. The compat mode provides a declarative mechanism (".pth" files that contain paths), but the newer-style ([PEP 660](https://peps.python.org/pep-0660/)) packages do not. There was some discussion about this a while back, but it ended without agreement on a way forward.

Running arbitrary untrusted Python code to resolve import locations is not the right answer, as that has significant security implications.
